// cgame/lib/ecs/components/Newtonian.js
var Vec2     = require('../../util/Vec2.js');
var Attached = require('./Attached.js');

module.exports = (function() {

  // Gives an entity information about its newtonian movement / state. It's
  // up to the chosen physics system to determine how to evolve this
  // information over time
  function Newtonian()
  {
    this.velocity     = Vec2.aquireOrClear(this.velocity);
    this.acceleration = Vec2.aquireOrClear(this.acceleration);
    this.radialSpeed = 0;

    // All scalars to limit magnitude of physics
    this.maxSpeed        = null;
    this.maxAcceleration = null;
    this.maxRadialSpeed  = null;
    this.radialRamp      = null;

    // relatives
    this._absVelocity = Vec2.aquireOrClear(this._absVelocity);
  }

  // In case of attached, sum the velocity vectors
  Newtonian.getAbsoluteVelocity = function(entity)
  {
    var v = entity.Newtonian._absVelocity;
    v.assign(entity.Newtonian.velocity);

    if (!entity.hasComponent(Attached)) {
      return v;
    }

    var parent = entity.Attached.parentEntity;
    return v.add(Newtonian.getAbsoluteVelocity(parent));
  };

  return Newtonian;

})();
