// cgame/index.js
module.exports = {

  Game  : require('./lib/Game.js'),
  Scene : require('./lib/Scene.js'),

  ecs: {
    Entity        : require('./lib/ecs/Entity.js'),
    EntityManager : require('./lib/ecs/EntityManager.js'),
    System        : require('./lib/ecs/System.js'),
    SystemManager : require('./lib/ecs/SystemManager.js'),

    systems: {
      Collider       : require('./lib/ecs/systems/Collider.js'),
      Expiry         : require('./lib/ecs/systems/Expiry.js'),
      Mover          : require('./lib/ecs/systems/Mover.js'),
      Physics        : require('./lib/ecs/systems/Physics.js'),
      Renderer       : require('./lib/ecs/systems/Renderer.js'),
      TreeMaintainer : require('./lib/ecs/systems/TreeMaintainer.js')
    },

    components: {
      Attached  : require('./lib/ecs/components/Attached.js'),
      Camera    : require('./lib/ecs/components/Camera.js'),
      Collision : require('./lib/ecs/components/Collision.js'),
      Elastic   : require('./lib/ecs/components/Elastic.js'),
      Expiry    : require('./lib/ecs/components/Expiry.js'),
      Newtonian : require('./lib/ecs/components/Newtonian.js'),
      Position  : require('./lib/ecs/components/Position.js'),
      Spatial   : require('./lib/ecs/components/Spatial.js'),
      Sprite    : require('./lib/ecs/components/Sprite.js')
    }
  },

  layers: {
    DebugLayer  : require('./lib/layers/DebugLayer.js'),
    EntityLayer : require('./lib/layers/EntityLayer.js'),
    FillLayer   : require('./lib/layers/FillLayer.js')
  }

};
