// cgame/Scene.js
module.exports = (function() {

  // Scene is NOT updating layers. Update function is still called by the
  // game, though
  Scene.PAUSED = 'paused';

  // Scene IS updating layers
  Scene.RUNNING = 'running';

  // Object containing layers to be updated and rendered to the screen.
  // Multiple scenes can be running at once
  function Scene()
  {
    this.layers = [];
    this.state = Scene.RUNNING;
  }

  // Add a new layers to this scene
  Scene.prototype.addLayer = function(layer)
  {
    this.layers.push(layer);
    this.sortLayers();
  };

  // Stop the updates
  Scene.prototype.pause = function()
  {
    this.state = Scene.PAUSED;
  };

  // Start playing again
  Scene.prototype.resume = function()
  {
    this.state = Scene.RUNNING;
  };

  // Sort layers
  var byZ = function(a, b) { return a.zIndex - b.zIndex; };
  Scene.prototype.sortLayers = function()
  {
    this.layers.sort(byZ);
  };

  // Fire off the update and draw methods for all layers
  Scene.prototype.update = function(dt, time)
  {
    if (this.state === Scene.RUNNING) {
      for (var n = 0; n < this.layers.length; n++) {
        var layer = this.layers[n];
        layer.update(dt, time);
      }
    }
  };

  return Scene;

})();
